### ISSUE TITLE
Running out of GPU Memory during training.

### ISSUE THREAD

**Issue by josefonte** [VLM-RL/Ocean-R1#3]
Hi,

I've tried running the script attached, in order to train a Qwen2.5VL-3B with the training data, but i kept coming across this error, indicating no more memory in gpu.

<img width="932" alt="Image" src="https://github.com/user-attachments/assets/10f92dde-5571-4e05-b23f-7d158a7f5465" />

i ran this with 4xA100 and with 4xH100, considering this is such a small model it should be more than enough. also i supervised the gpu utilization during training, and it fluctuated a lot with very high volatility at times (which i would assume it's normal but pls correct me if i'm wrong). 

Is this a known issue, and is there solution? 

thanks in advance :) 

Note:
- I had to update verl directory to the most recent version of their github repo, due to incompatibility with vllm version and Qwen model. 
- using updated version of the script to be able to run (see attachement).
- also attached the list of modules and libs in conda env, and their respective versions.

[reqs.txt](https://github.com/user-attachments/files/19708028/reqs.txt)
[run_qwen25vl-3b_stage1.txt](https://github.com/user-attachments/files/19708047/run_qwen25vl-3b_stage1.txt)

**Comment by fengzi258** [comment]
It is normal for the GPU utilization rate to fluctuate greatly because of the  `rollout` process.

You meet OOM Error, you can try reduce the following hyperparameters:
- `data.train_batch_size`
- `actor_rollout_ref.actor.ppo_mini_batch_size`
- `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu`
- `actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu`
- `actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu`
- `actor_rollout_ref.rollout.n`

Upload the running script can provide more informations.

**Comment by josefonte** [comment]
hi, 

I experienced a lot of issues running the original script provided (due to ray configuration) so i got the scripts provided in verl repo [link](https://github.com/volcengine/verl/blob/main/examples/grpo_trainer/run_qwen2_5_vl-7b.sh) that train qwen2.5 VL 7b and changed it to use Ocean-R1 datasets (stage1).  FYI: I ran the script provided and it trained perfectly fine with Geo3k.

this is my script right now after several tweaks taking ur sugestions, trying to reduce the amount of memory in GPU .  any further sugestions? 

are there any params that i shouldn't change at all? like `max_prompt_length` or `max_num_batched_tokens`.

is the error bc of the size of the dataset? I thought about saving the images and replacing the column with images bytes with a path for a folder with each image. would that help in any way?

```
PYTHONUNBUFFERED=1 python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=$HOME/Ocean-R1/train_data/minglingfeng/Ocean_R1_visual_data_stage1.parquet \
    data.val_files="['$HOME/Ocean-R1/eval_data/minglingfeng/cvbench_test.parquet','$HOME/Ocean-R1/eval_data/minglingfeng/geoqa_test.parquet']" \
    data.train_batch_size=16 \
    data.max_prompt_length=6500 \
    data.max_response_length=4096 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.image_key=images \
    actor_rollout_ref.model.path=$HOME/models/Qwen2.5-VL-3B-Instruct \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.ppo_mini_batch_size=8 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.01 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.n=2 \
    actor_rollout_ref.rollout.max_num_batched_tokens=2048 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=2\
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.enable_chunked_prefill=False \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=False \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger=['console'] \
    trainer.project_name='ocean-r1-grpo-stage1' \
    trainer.experiment_name='ocean-r1-grpo-stage1' \
    trainer.n_gpus_per_node=4 \
    trainer.nnodes=1 \
    trainer.save_freq=-1 \
    trainer.test_freq=5 \
    trainer.total_epochs=15 "$@"
```

**Comment by fengzi258** [comment]
> hi,
> 
> I experienced a lot of issues running the original script provided (due to ray configuration) so i got the scripts provided in verl repo [link](https://github.com/volcengine/verl/blob/main/examples/grpo_trainer/run_qwen2_5_vl-7b.sh) that train qwen2.5 VL 7b and changed it to use Ocean-R1 datasets (stage1). FYI: I ran the script provided and it trained perfectly fine with Geo3k.
> 
> this is my script right now after several tweaks taking ur sugestions, trying to reduce the amount of memory in GPU . any further sugestions?
> 
> are there any params that i shouldn't change at all? like `max_prompt_length` or `max_num_batched_tokens`.
> 
> is the error bc of the size of the dataset? I thought about saving the images and replacing the column with images bytes with a path for a folder with each image. would that help in any way?
> 
> ```
> PYTHONUNBUFFERED=1 python3 -m verl.trainer.main_ppo \
>     algorithm.adv_estimator=grpo \
>     data.train_files=$HOME/Ocean-R1/train_data/minglingfeng/Ocean_R1_visual_data_stage1.parquet \
>     data.val_files="['$HOME/Ocean-R1/eval_data/minglingfeng/cvbench_test.parquet','$HOME/Ocean-R1/eval_data/minglingfeng/geoqa_test.parquet']" \
>     data.train_batch_size=16 \
>     data.max_prompt_length=6500 \
>     data.max_response_length=4096 \
>     data.filter_overlong_prompts=True \
>     data.truncation='error' \
>     data.image_key=images \
>     actor_rollout_ref.model.path=$HOME/models/Qwen2.5-VL-3B-Instruct \
>     actor_rollout_ref.model.use_remove_padding=True \
>     actor_rollout_ref.model.enable_gradient_checkpointing=True \
>     actor_rollout_ref.actor.optim.lr=1e-6 \
>     actor_rollout_ref.actor.ppo_mini_batch_size=8 \
>     actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=4 \
>     actor_rollout_ref.actor.use_kl_loss=True \
>     actor_rollout_ref.actor.kl_loss_coef=0.01 \
>     actor_rollout_ref.actor.kl_loss_type=low_var_kl \
>     actor_rollout_ref.actor.entropy_coeff=0 \
>     actor_rollout_ref.actor.fsdp_config.param_offload=True \
>     actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
>     actor_rollout_ref.rollout.name=vllm \
>     actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
>     actor_rollout_ref.rollout.n=2 \
>     actor_rollout_ref.rollout.max_num_batched_tokens=2048 \
>     actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=2\
>     actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
>     actor_rollout_ref.rollout.enable_chunked_prefill=False \
>     actor_rollout_ref.rollout.enforce_eager=False \
>     actor_rollout_ref.rollout.free_cache_engine=False \
>     actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
>     actor_rollout_ref.ref.fsdp_config.param_offload=True \
>     algorithm.use_kl_in_reward=False \
>     trainer.critic_warmup=0 \
>     trainer.logger=['console'] \
>     trainer.project_name='ocean-r1-grpo-stage1' \
>     trainer.experiment_name='ocean-r1-grpo-stage1' \
>     trainer.n_gpus_per_node=4 \
>     trainer.nnodes=1 \
>     trainer.save_freq=-1 \
>     trainer.test_freq=5 \
>     trainer.total_epochs=15 "$@"
> ```

Notably, the method of saving and parsing images corresponds to the code in `.verl/verl/utils/dataset/rl_dataset.py` (We add the parsing `images bytes` in our code). 
you can refer to [issues/2](https://github.com/VLM-RL/Ocean-R1/issues/2) for details. Check if it is correct to parse the image in your code? 
If u still meet OOM Error, u may sample the sub-dataset of  `Ocean-R1 datasets `, e.g, take the first 10000 samples.

**Comment by josefonte** [comment]
I've tried doing ur recomendation of using a subset of the dataset (100samples) and ran this script 

```
PYTHONUNBUFFERED=1 python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=$HOME/Ocean-R1/train_data/minglingfeng/Ocean_R1_visual_data_stage1_100.parquet \
    data.val_files=$HOME/Ocean-R1/eval_data/minglingfeng/cvbench_test.parquet \
    data.train_batch_size=4 \
    data.max_prompt_length=8192 \
    data.max_response_length=4096 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.image_key=images \
    actor_rollout_ref.model.path=$HOME/models/Qwen2.5-VL-3B-Instruct \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.ppo_mini_batch_size=4 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=2 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.01 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.tensor_model_parallel_size=4 \
    actor_rollout_ref.rollout.n=2 \
    actor_rollout_ref.rollout.max_num_batched_tokens=4096 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=2\
    actor_rollout_ref.rollout.gpu_memory_utilization=0.9 \
    actor_rollout_ref.rollout.enable_chunked_prefill=False \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=False \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger=['console'] \
    trainer.project_name='ocean-r1-grpo-stage1-10k' \
    trainer.experiment_name='ocean-r1-grpo-stage1-10k' \
    trainer.n_gpus_per_node=4 \
    trainer.nnodes=1 \
    trainer.save_freq=20 \
    trainer.test_freq=5 \
    trainer.total_epochs=15 "$@"

```

**Comment by fengzi258** [comment]
> I've tried doing ur recomendation of using a subset of the dataset (100samples) and ran this script
> 
> ```
> PYTHONUNBUFFERED=1 python3 -m verl.trainer.main_ppo \
>     algorithm.adv_estimator=grpo \
>     data.train_files=$HOME/Ocean-R1/train_data/minglingfeng/Ocean_R1_visual_data_stage1_100.parquet \
>     data.val_files=$HOME/Ocean-R1/eval_data/minglingfeng/cvbench_test.parquet \
>     data.train_batch_size=4 \
>     data.max_prompt_length=8192 \
>     data.max_response_length=4096 \
>     data.filter_overlong_prompts=True \
>     data.truncation='error' \
>     data.image_key=images \
>     actor_rollout_ref.model.path=$HOME/models/Qwen2.5-VL-3B-Instruct \
>     actor_rollout_ref.model.use_remove_padding=True \
>     actor_rollout_ref.model.enable_gradient_checkpointing=True \
>     actor_rollout_ref.actor.optim.lr=1e-6 \
>     actor_rollout_ref.actor.ppo_mini_batch_size=4 \
>     actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=2 \
>     actor_rollout_ref.actor.use_kl_loss=True \
>     actor_rollout_ref.actor.kl_loss_coef=0.01 \
>     actor_rollout_ref.actor.kl_loss_type=low_var_kl \
>     actor_rollout_ref.actor.entropy_coeff=0 \
>     actor_rollout_ref.actor.fsdp_config.param_offload=True \
>     actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
>     actor_rollout_ref.rollout.name=vllm \
>     actor_rollout_ref.rollout.tensor_model_parallel_size=4 \
>     actor_rollout_ref.rollout.n=2 \
>     actor_rollout_ref.rollout.max_num_batched_tokens=4096 \
>     actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=2\
>     actor_rollout_ref.rollout.gpu_memory_utilization=0.9 \
>     actor_rollout_ref.rollout.enable_chunked_prefill=False \
>     actor_rollout_ref.rollout.enforce_eager=False \
>     actor_rollout_ref.rollout.free_cache_engine=False \
>     actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
>     actor_rollout_ref.ref.fsdp_config.param_offload=True \
>     algorithm.use_kl_in_reward=False \
>     trainer.critic_warmup=0 \
>     trainer.logger=['console'] \
>     trainer.project_name='ocean-r1-grpo-stage1-10k' \
>     trainer.experiment_name='ocean-r1-grpo-stage1-10k' \
>     trainer.n_gpus_per_node=4 \
>     trainer.nnodes=1 \
>     trainer.save_freq=20 \
>     trainer.test_freq=5 \
>     trainer.total_epochs=15 "$@"
> ```

what's the problem now? still OOM?

**Comment by josefonte** [comment]
yes :(

**Comment by josefonte** [comment]
yes :( 

I'm running the script using 4xA100

<img width="1274" alt="Image" src="https://github.com/user-attachments/assets/0d097653-f141-4749-991b-f2c867c82d3a" />

**Comment by josefonte** [comment]
Hi, 

I tried doing a filtering of the data, filtering high dim images (threshold of 500000 px was the same as the max dim of the other dataset i tested and it ran). the dataset went from 63k to 60k. tried it and failed. did a subsample of 1k images of the small dataset, and it also failed. could there be a memory leak of some sorts. 

also could u provide more info regarding versions of ur environment u used to train (f.e : cuda, python, torch, vllm, flash)? 

rn, I download the most recent version of verl repo and updated the env to use torch 2.6 on cuda 12.4, vllm 0.8.3 ,flash-attn 2.7.4.post1
